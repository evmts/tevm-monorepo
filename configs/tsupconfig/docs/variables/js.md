**@tevm/tsupconfig** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > js

# Variable: js

> **`const`** **js**: `Options`

## Source

[js.js:3](https://github.com/evmts/tevm-monorepo/blob/main/configs/tsupconfig/src/js.js#L3)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
