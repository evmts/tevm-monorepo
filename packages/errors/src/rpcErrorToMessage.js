/**
 * @type {Record<import('./requestErrors.js').RpcErrorCode, string>}
 */
export const rpcErrorCodeToMessage = {
	[-1]: 'Unknwon error',
	[-32700]: 'Parse error',
	[-32600]: 'Invalid request',
	[-32601]: 'Method not found',
	[-32602]: 'Invalid params',
	[-32603]: 'Internal error',
	[-32000]: 'Invalid input',
	[-32001]: 'Resource not found',
	[-32002]: 'Resource unavailable',
	[-32003]: 'Transaction rejected',
	[-32004]: 'Method not supported',
	[-32005]: 'Limit exceeded',
	[-32006]: 'JSON-RPC version not supported',
	[-32042]: 'Method not found',
}
