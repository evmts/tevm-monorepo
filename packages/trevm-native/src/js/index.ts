import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};