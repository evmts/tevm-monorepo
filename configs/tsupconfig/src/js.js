import { createTsUpOptions } from './createTsupOptions.js'

export const js = createTsUpOptions({
	target: 'js',
})
