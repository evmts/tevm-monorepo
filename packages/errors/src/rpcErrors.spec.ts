import { describe, expect, it } from 'vitest'
import {
	InternalRpcError,
	InvalidInputRpcError,
	InvalidParamsRpcError,
	InvalidRequestRpcError,
	JsonRpcVersionUnsupportedError,
	LimitExceededRpcError,
	MethodNotFoundRpcError,
	MethodNotSupportedRpcError,
	ParseRpcError,
	ProviderDisconnectedError,
	ResourceNotFoundRpcError,
	ResourceUnavailableRpcError,
	TimeoutError,
	UnknownRpcError,
	UnsupportedProviderMethodError,
	UserRejectedRequestError,
} from './requestErrors.js'
import { rpcErrorCodeToMessage } from './rpcErrorToMessage.js'

describe('RPC Error Classes', () => {
	it('should have a valid rpcErrorCodeToMessage mapping', () => {
		expect(rpcErrorCodeToMessage).toMatchSnapshot('rpcErrorCodeToMessage')
	})

	it('should create proper instances of RPC errors', () => {
		// Test each RPC error class with snapshot tests
		const rpcErrors = [
			new InternalRpcError(new Error('Internal error occurred')),
			new InvalidInputRpcError(new Error('Invalid input provided')),
			new InvalidParamsRpcError(new Error('Invalid parameters')),
			new InvalidRequestRpcError(new Error('Invalid request')),
			new JsonRpcVersionUnsupportedError(new Error('JSON-RPC version not supported')),
			new LimitExceededRpcError(new Error('Limit exceeded')),
			new MethodNotFoundRpcError(new Error('Method not found')),
			new MethodNotSupportedRpcError(new Error('Method not supported')),
			new ParseRpcError(new Error('Parse error')),
			// ProviderDisconnectedError expects a string argument
			new ProviderDisconnectedError('Provider disconnected' as any),
			new ResourceNotFoundRpcError(new Error('Resource not found')),
			new ResourceUnavailableRpcError(new Error('Resource unavailable')),
			new UnknownRpcError(new Error('Unknown error')),
			// UnsupportedProviderMethodError accepts an object with method property
			new UnsupportedProviderMethodError({ method: 'eth_unsupportedMethod' } as any),
			new UserRejectedRequestError(new Error('User rejected request')),
			// TimeoutError expects request parameters
			new TimeoutError({ url: 'https://example.com', body: {} } as any),
		]

		// Test each error instance with a snapshot
		rpcErrors.forEach((error) => {
			expect(error).toMatchSnapshot(error.constructor.name)
		})
	})
})
