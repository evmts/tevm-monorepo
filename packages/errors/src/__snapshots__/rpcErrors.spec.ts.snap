// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RPC Error Classes > should create proper instances of RPC errors > InternalRpcError 1`] = `
[InternalRpcError: An internal error was received.

Details: Internal error occurred
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > InvalidInputRpcError 1`] = `
[InvalidInputRpcError: Missing or invalid parameters.
Double check you have provided the correct parameters.

Details: Invalid input provided
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > InvalidParamsRpcError 1`] = `
[InvalidParamsRpcError: Invalid parameters were provided to the RPC method.
Double check you have provided the correct parameters.

Details: Invalid parameters
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > InvalidRequestRpcError 1`] = `
[InvalidRequestRpcError: JSON is not a valid request object.

Details: Invalid request
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > JsonRpcVersionUnsupportedError 1`] = `
[JsonRpcVersionUnsupportedError: Version of JSON-RPC protocol is not supported.

Details: JSON-RPC version not supported
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > LimitExceededRpcError 1`] = `
[LimitExceededRpcError: Request exceeds defined limit.

Details: Limit exceeded
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > MethodNotFoundRpcError 1`] = `
[MethodNotFoundRpcError: The method does not exist / is not available.

Details: Method not found
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > MethodNotSupportedRpcError 1`] = `
[MethodNotSupportedRpcError: Method is not supported.

Details: Method not supported
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > ParseRpcError 1`] = `
[ParseRpcError: Invalid JSON was received by the server. An error occurred on the server while parsing the JSON text.

Details: Parse error
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > ProviderDisconnectedError 1`] = `
[ProviderDisconnectedError: The Provider is disconnected from all chains.

Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > ResourceNotFoundRpcError 1`] = `
[ResourceNotFoundRpcError: Requested resource not found.

Details: Resource not found
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > ResourceUnavailableRpcError 1`] = `
[ResourceUnavailableRpcError: Requested resource not available.

Details: Resource unavailable
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > TimeoutError 1`] = `
[TimeoutError: The request took too long to respond.

URL: https://example.com
Request body: {}

Details: The request timed out.
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > UnknownRpcError 1`] = `
[UnknownRpcError: An unknown RPC error occurred.

Details: Unknown error
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > UnsupportedProviderMethodError 1`] = `
[UnsupportedProviderMethodError: The Provider does not support the requested method.

Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should create proper instances of RPC errors > UserRejectedRequestError 1`] = `
[UserRejectedRequestError: User rejected the request.

Details: User rejected request
Version: viem@2.23.10]
`;

exports[`RPC Error Classes > should have a valid rpcErrorCodeToMessage mapping > rpcErrorCodeToMessage 1`] = `
{
  "-1": "Unknwon error",
  "-32000": "Invalid input",
  "-32001": "Resource not found",
  "-32002": "Resource unavailable",
  "-32003": "Transaction rejected",
  "-32004": "Method not supported",
  "-32005": "Limit exceeded",
  "-32006": "JSON-RPC version not supported",
  "-32042": "Method not found",
  "-32600": "Invalid request",
  "-32601": "Method not found",
  "-32602": "Invalid params",
  "-32603": "Internal error",
  "-32700": "Parse error",
}
`;
